File: hard/2872. Maximum Number of K-Divisible Components/solution.java
class Solution {
    public int maxKDivisibleComponents(int n, int[][] edges, int[] values, int k) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            graph.get(edge[0]).add(edge[1]);
            graph.get(edge[1]).add(edge[0]);
        }
        int[] components = {0};
        dfs(0, -1, graph, values, k, components);
        return components[0];
    }

    private long dfs(int node, int parent, List<List<Integer>> graph, int[] values, int k, int[] components) {
        long sum = values[node];
        for (int neighbor : graph.get(node)) {
            if (neighbor != parent) {
                sum += dfs(neighbor, node, graph, values, k, components);
            }
        }
        if (sum % k == 0) {
            components[0]++;
            return 0;
        }
        return sum;
    }
}

File: hard/2872. Maximum Number of K-Divisible Components/solution.rs
impl Solution {
    pub fn max_k_divisible_components(n: i32, edges: Vec<Vec<i32>>, values: Vec<i32>, k: i32) -> i32 {
        let n = n as usize;
        let k = k as i64;
        let mut graph = vec![vec![]; n];
        for edge in edges {
            let u = edge[0] as usize;
            let v = edge[1] as usize;
            graph[u].push(v);
            graph[v].push(u);
        }
        
        let mut components = 0;
        Self::dfs(0, n, &graph, &values, k, &mut components);
        components
    }
    
    fn dfs(node: usize, parent: usize, graph: &Vec<Vec<usize>>, values: &Vec<i32>, k: i64, components: &mut i32) -> i64 {
        let mut sum = values[node] as i64;
        for &neighbor in &graph[node] {
            if neighbor != parent {
                sum += Self::dfs(neighbor, node, graph, values, k, components);
            }
        }
        if sum % k == 0 {
            *components += 1;
            return 0;
        }
        sum
    }
}